Add feedback upload in CSV score files

mwest1066 opened the pull request:
Adds the capability to set feedback when uploading instance question scores by CSV.

Closes #2049
Closes #1384

coatless commented:
```
mwest@illinois.edu,1,addTwoNumbers,34.5,42983,"{""manual"": ""You _divided_ rather than _multiplying_""}"
```

Anyway to drop the `{"manual":}` component? TAs are likely to add feedback via CSV.

nicknytko commented on lib/score-upload.js:
```diff
@@ -291,29 +301,22 @@ module.exports = {
-                });
-            } else {
+            if (!_.has(json, 'score_perc') && !_.has(json, 'points')) {
                 return callback(new Error('must specify either "score_perc" or "points"'));
```
Nice, this looks a lot more clean than before

bowmnath commented:
Looks great Matt!

I would second the comment about removing the JSON aspect from the input if possible.  I don't think the end user gains anything from it (unless you designed it for some use case I haven't thought of), and I imagine it would be frustrating for anyone who had to type that over and over.  On the other hand, I (and almost every PL user I know) would avoid that with a simple Python script after editing the CSV, so it's certainly not the end of the world!

mwest1066 commented:
@coatless @bowmanat I changed the default for the `feedback` column to be just a plain string (which can be rendered as Markdown). I kept the full JSON functionality with an optional `feedback_json` column.

I also added two new Downloads files, one CSV that has all the information needed for grading and columns ready for the score and feedback, and a ZIP of any corresponding files for manual grading.

There are new docs about manual grading, copied below.

#############################################################

# Manual Grading

PrairieLearn supports manual grading of questions by downloading a CSV file with student answers and uploading a CSV file with question scores and optional per-question feedback. There is not currently an online web interface for manual grading.


## Configuring a question for maual grading

The [`info.json` file](question.md#question-infojson) in the question should be set for manual grading:
```json
{
    "uuid": "cbf5cbf2-6458-4f13-a418-aa4d2b1093ff",
    "gradingMethod": "Manual",
    "singleVariant": true,
    ...
}
```

Manually-graded questions allow students to "Save" answers, but they don't have a "Save & Grade" button. Instead, the student just saves answers as many times as they want, and all of their submitted answers are stored. It is recommended to also mark manually-graded questions as `"singleVariant": true` so that students are only given a single random variant, even on Homework assessments.

Any [elements](elements/) can be used in the [`question.html`](question.md#question-questionhtml) to write manually graded questions. All of the student input will be saved and available for manual grading, including `pl-string-input`, `pl-file-editor`, `pl-file-upload`, etc.

To show manual feedback the `question.html` file should contain an element to display the feedback next to student submissions. A basic template for this is:
```html
<pl-submission-panel>
  {{#feedback.manual}}
  <p>Feedback from course staff:</p>
  <markdown>{{{feedback.manual}}}</markdown>
  {{/feedback.manual}}
</pl-submission-panel>
```

This example template formats the feedback as Markdown.


## Downloading the students' submitted answers

After students have completed the assessment, download the submitted answers by going to the assessment page, then the "Downloads" tab, and selecting the `<assessment>_submissions_for_manual_grading.csv` file. This looks like:
```csv
uid,qid,submission_id,params,true_answer,submitted_answer,score_perc,feedback
mwest@illinois.edu,explainMax,42983,{},{},{"ans": "returns the maximum value in the array"},,
zilles@illinois.edu,explainMax,42984,{},{},{"ans": "gives the set of largest values in the object"},,
zilles@illinois.edu,describeFibonacci,42987,{},{},{"ans": "calculates the n-th Fibonacci number"},,
```

This CSV file has two blank columns at the end, ready for the percentage score and optional feedback. The `submission_id` is an internal identifier that PrairieLearn uses to determine exactly which submitted answer is being graded. The `params` and `true_answer` columns show the question data.

If the students uploaded files then you should also download `<assessment>_files_for_manual_grading.zip` from the Downloads page. The scores and feedback should still be entered into the CSV file.


## Uploading the scores and feedback

After editing the percentage score and/or feedback for each submitted answer, upload the CSV file by going to the assessment page, then the "Uploads" tab, and selecting "Upload new question scores".

Each question will have its score and/or feedback updated and the total assessment score will be recalculated. All updates are done with `credit` of 100%, so students get exactly the scores as uploaded.

If you prefer to use points rather than a percentage score, rename the `score_perc` column in the CSV file to `points`.

coatless commented on doc/manualGrading.md:
````diff
@@ -0,0 +1,56 @@
+}
+```
+
+Manually-graded questions allow students to "Save" answers, but they don't have a "Save & Grade" button. Instead, the student just saves answers as many times as they want, and all of their submitted answers are stored. It is recommended to also mark manually-graded questions as `"singleVariant": true` so that students are only given a single random variant, even on Homework assessments.
````
Emphasize the `"gradingMethod": "Manual"` key?

mwest1066 replied:
done

coatless commented on doc/manualGrading.md:
````diff
@@ -0,0 +1,56 @@
+
+To show manual feedback the `question.html` file should contain an element to display the feedback next to student submissions. A basic template for this is:
+```html
+<pl-submission-panel>
````
Any reason for not recommending it be placed in `pl-answer-panel`?

mwest1066 replied:
Feedback is specially associated with a submission, because it's a comment on the submission itself. The "answer" panel is supposed to be for the "gold standard answer".

coatless commented on doc/manualGrading.md:
````diff
@@ -0,0 +1,56 @@
+zilles@illinois.edu,describeFibonacci,42987,{},{},{"ans": "calculates the n-th Fibonacci number"},,
+```
+
+This CSV file has two blank columns at the end, ready for the percentage score and optional feedback. The `submission_id` is an internal identifier that PrairieLearn uses to determine exactly which submitted answer is being graded. The `params` and `true_answer` columns show the question data.
````
Instead of empty, which could be problematic.... How about:

1. Set all `score_perc` to 0 in the default CSV. 
2. Set a common message in `feedback` to `"Fill me in!"` or `"Great job on this question!"`

The latter can be dynamically changed per user with random adjectives if we want to mirror [`praise`](https://github.com/rladies/praise).

mwest1066 replied:
Blank `score_perc` fields just don't set the score for that student, and blank `feedback` fields don't set feedback. I thought this was useful so that you could upload partially-complete grading CSV files?

I could add an extra column called `old_score_perc`, containing the current score? That way people could copy the old score if they are doing a regrade. On a first-ever grading the `old_score_perc` would be zero for all questions.

bowmnath replied:
For what it's worth, I prefer the blank fields.  For me, putting in an explicit 0 is different from not having graded it yet.  Updating a single grade is already pretty easy by just submitting a CSV with just that student so I can be absolutely sure I'm not messing up someone I already graded.

mwest1066 replied:
I added the `old_score_perc` column, but I left the new columns blank.

mwest1066 replied:
I also clarified in the docs that if you leave either `score_perc` or `feedback` (or both) blank for any student, then the corresponding entry will not be updated.

coatless commented on pages/instructorAssessmentUploads/csvHelpInstanceQuestionScores.ejs:
```diff
@@ -1,25 +1,19 @@
 <p>
-  Upload a CSV file like this:
+  Upload a CSV file in the format of
```
Perhaps add a link out to the documentation for manual grading?

mwest1066 replied:
done

mwest1066 commented:
I added a news item:

[PrairieLearn.pdf](https://github.com/PrairieLearn/PrairieLearn/files/4350449/PrairieLearn.pdf)

@coatless @daviddalpiaz @bowmnath please check if you get a chance!

bowmnath commented:
News item looks good to me!  I didn't have any "But how do I..." pop up while reading it.  Seems really straightforward.
